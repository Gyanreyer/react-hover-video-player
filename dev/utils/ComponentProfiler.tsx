/* eslint-disable no-console */
import React from 'react';

interface RenderTiming {
  averageRenderTime: number;
  renderCount: number;
}

interface ComponentProfilerProps {
  profilerID: string;
  children: React.ReactChild;
}

/**
 * Util component logs out render timings for whatever component it's wrapped aroud
 */
const ComponentProfiler = ({
  profilerID,
  children,
}: ComponentProfilerProps): JSX.Element => {
  // Use a ref to track render timing for this component in a persistent store
  const renderTiming = React.useRef<RenderTiming>();
  if (!renderTiming.current) {
    renderTiming.current = {
      averageRenderTime: 0,
      renderCount: 0,
    };
  }

  // Logs out helpful render timing info for performance measurements
  const onProfilerRender = React.useCallback(
    (
      id, // the "id" prop of the Profiler tree that has just committed
      phase: string, // either "mount" (if the tree just mounted) or "update" (if it re-rendered)
      actualDuration: number // time spent rendering the committed update
    ) => {
      if (phase === 'mount') {
        console.log(`${profilerID} | MOUNT: ${actualDuration}ms`);
      } else {
        renderTiming.current.renderCount += 1;
        renderTiming.current.averageRenderTime +=
          (actualDuration - renderTiming.current.averageRenderTime) /
          renderTiming.current.renderCount;
        console.log(
          `${profilerID} | UPDATE: ${actualDuration}ms | New average: ${renderTiming.current.averageRenderTime}ms`
        );
      }
    },
    [profilerID]
  );
  return (
    <React.Profiler id={profilerID} onRender={onProfilerRender}>
      {children}
    </React.Profiler>
  );
};

export default ComponentProfiler;
