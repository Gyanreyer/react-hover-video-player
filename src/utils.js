/**
 * @typedef   {object}  VideoSource
 * @property  {string}  src - The src URL string to use for a video player source
 * @property  {string}  type - The media type of the video, ie 'video/mp4'
 */

/**
 * @function  formatVideoSrc
 *
 * Takes a videoSrc value and formats it as an array of VideoSource objects which can be used to render
 * <source> elements for the video
 *
 * @param {!(string|string[]|VideoSource|VideoSource[])}  videoSrc - Source(s) to use for the video player. Accepts 3 different formats:
 *                                                                   - **String**: the URL string to use as the video player's src
 *                                                                   - **Object**: an object with attributes:
 *                                                                     - src: The src URL string to use for a video player source
 *                                                                     - type: The media type of the video source, ie 'video/mp4'
 *                                                                   - **Array**: if you would like to provide multiple sources, you can provide an array of URL strings and/or objects with the shape described above
 *
 * @returns {VideoSource[]} Array of formatted VideoSource objects which can be used to render <source> elements for the video
 */
export function formatVideoSrc(videoSrc) {
  if (videoSrc == null) {
    // A videoSrc value is required in order to make the video player work
    console.error(
      "Error: 'videoSrc' prop is required for HoverVideoPlayer component"
    );

    return [];
  }

  return (
    // Make sure we can treat the videoSrc value as an array
    []
      .concat(videoSrc)
      // Parse our video source values into an array of VideoSource objects that can be used to render sources for the video
      .reduce((sourceArray, source) => {
        if (typeof source === 'string') {
          // If the source is a string, it's an src URL so format it into a VideoSource object and add it to the array
          sourceArray.push({ src: source });
        } else if (source && source.src) {
          // If the source is an object with an src, just add it to the array
          sourceArray.push({ src: source.src, type: source.type });
        } else {
          // Log an error if one of the videoSrc values is invalid
          console.error(
            "Error: invalid value provided to HoverVideoPlayer prop 'videoSrc':",
            source
          );
        }

        return sourceArray;
      }, [])
  );
}

/**
 * @typedef   {object}  VideoCaptionsTrack
 * @property  {string}  src - The src URL string for the captions track file
 * @property  {string}  srcLang - The language code for the language that these captions are in
 * @property  {string}  label - The title of the captions track
 * @property  {bool}    default - Whether this track should be used by default if the user's preferences don't match an available srcLang
 */

/**
 * @function formatVideoCaptions
 *
 * Takes a videoCaptions value and formats it as an array of VideoCaptionsTrack objects which can be used to render
 * <track> elements for the video
 *
 * @param {!(string|string[]|VideoCaptionsTrack|VideoCaptionsTrack[])} videoCaptions - Captions track(s) to use for the video player for accessibility. Accepts 3 different formats:
 *                                                                                     - **String**: the URL string to use as the captions track's src
 *                                                                                     - **Object**: an object with attributes:
 *                                                                                       - src: The src URL string for the captions track file
 *                                                                                       - srcLang: The language code for the language that these captions are in (ie, 'en', 'es', 'fr')
 *                                                                                       - label: The title of the captions track
 *                                                                                       - default: Whether this track should be used by default if the user's preferences don't match an available srcLang
 *                                                                                     - **Array**: if you would like to provide multiple caption tracks, you can provide an array of objects with the shape described above
 *
 * @returns {VideoCaptionsTrack[]}  Array of formatted VideoCaptionsTrack objects which can be used to render <track> elements for the video
 */
export function formatVideoCaptions(videoCaptions) {
  // If no captions were provided, return an empty array
  if (videoCaptions == null) return [];

  return (
    // Make sure we can treat the videoSrc value as an array
    []
      .concat(videoCaptions)
      // Parse our video captions values into an array of VideoCaptionsTrack
      // objects that can be used to render caption tracks for the video
      .reduce((captionsArray, captions) => {
        if (captions && captions.src) {
          captionsArray.push({
            src: captions.src,
            srcLang: captions.srcLang,
            label: captions.label,
            default: Boolean(captions.default),
          });
        } else {
          // Log an error if one of the videoCaptions values is invalid
          console.error(
            "Error: invalid value provided to HoverVideoPlayer prop 'videoCaptions'",
            captions
          );
        }

        return captionsArray;
      }, [])
  );
}
