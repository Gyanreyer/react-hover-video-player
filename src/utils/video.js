/* eslint-disable no-param-reassign */
// Enumerates states that the video can be in
export const VIDEO_STATE = {
  paused: 'paused',
  loading: 'loading',
  playing: 'playing',
};

/**
 * @function getVideoState
 *
 * Takes a video element and returns its current playing state
 *
 * @param {node} videoElement
 */
export function getVideoState(videoElement) {
  if (videoElement.paused || videoElement.ended) {
    return VIDEO_STATE.paused;
  }

  // If the video isn't paused but its readyState indicates it isn't loaded enough
  // to play yet, it is loading
  if (videoElement.readyState < 3) {
    return VIDEO_STATE.loading;
  }

  // If the video isn't paused and its ready state indicates it's loaded enough to play,
  // assume it's playing
  return VIDEO_STATE.playing;
}

/**
 * @function playVideo
 *
 * Takes a video element and attempts to play it.
 *
 * @param {node}  videoElement - The video element to start playing
 *
 * @returns {Promise}   A Promise which will resolve when the video starts playing or reject if an error occurs
 */
export function playVideo(videoElement) {
  // Start playing the video and hang onto the play promise it returns
  let playPromise = videoElement.play();

  if (!playPromise || !playPromise.then) {
    // If videoElement.play() didn't return a promise, we'll manually create one
    // ourselves which mimics the same behavior
    playPromise = new Promise((resolve, reject) => {
      // Declaring onVideoPlaybackFailed up here so we can refer to it and remove its event listener
      // if the video successfully starts playing
      let onVideoPlaybackFailed;

      // Set up event listener to resolve the promise when the video player starts playing
      const onVideoPlaybackStarted = () => {
        // Remove the event listeners we added as cleanup now that the play attempt has succeeded
        videoElement.removeEventListener('playing', onVideoPlaybackStarted);
        videoElement.removeEventListener('error', onVideoPlaybackFailed);

        // Resolve because we successfully started playing!
        resolve();
      };
      videoElement.addEventListener('playing', onVideoPlaybackStarted, {
        once: true,
      });

      // Set up event listener to reject the promise when the video player encounters an error
      onVideoPlaybackFailed = (event) => {
        // Remove the event listeners we added as cleanup now that the play attempt has failed
        videoElement.removeEventListener('error', onVideoPlaybackFailed);
        videoElement.removeEventListener('playing', onVideoPlaybackStarted);

        // Reject with the error that was thrown
        reject(event.error);
      };
      videoElement.addEventListener('error', onVideoPlaybackFailed);
    });
  }

  return playPromise;
}

/**
 * @function  pauseVideo
 *
 * Pauses the video and resets it to the beginning if necessary
 *
 * @param {node}  videoElement - The video element to pause playback for
 * @param {bool}  shouldRestart - Whether the videp should be reset to the beginning after pausing
 */
export function pauseVideo(videoElement, shouldRestart) {
  videoElement.pause();

  if (shouldRestart) {
    // If we should restart the video, reset its time to the beginning
    videoElement.currentTime = 0;
  }
}
