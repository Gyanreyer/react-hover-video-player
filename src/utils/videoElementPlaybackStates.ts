/**
 * Takes a video element and returns whether it is currently paused, meaning it is not
 * playing or attempting to play.
 *
 * @param {HTMLVideoElement} videoElement
 */
export function isVideoElementPaused(
  videoElement: HTMLVideoElement | null
): boolean {
  return videoElement ? videoElement.paused || videoElement.ended : true;
}
