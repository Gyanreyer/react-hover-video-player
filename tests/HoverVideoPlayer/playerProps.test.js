import React from 'react';
import { fireEvent, act } from '@testing-library/react';

import { renderHoverVideoPlayer, getPlayPromise } from '../utils';

describe('Video props', () => {
  test('isVideoMuted prop correctly sets muted attribute on video', () => {
    const { container, rerenderWithProps } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
    });

    const videoElement = container.querySelector('video');

    // muted should be true by default
    expect(videoElement.muted).toBe(true);

    // Re-render with the video unmuted
    rerenderWithProps({ videoSrc: 'fake/video-file.mp4', muted: false });
    expect(videoElement.muted).toBe(false);

    // Re-render with muted explicitly set to true
    rerenderWithProps({ videoSrc: 'fake/video-file.mp4', muted: true });
    expect(videoElement.muted).toBe(true);
  });

  test('shouldVideoLoop prop correctly sets loop attribute on video', () => {
    const { container, rerenderWithProps } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
    });

    const videoElement = container.querySelector('video');

    // Loop should be true by default
    expect(videoElement).toHaveAttribute('loop');

    // Re-render with looping disabled on the video
    rerenderWithProps({
      videoSrc: 'fake/video-file.mp4',
      loop: false,
    });
    expect(videoElement).not.toHaveAttribute('loop');

    // Re-render with loop explicitly set to true on the video
    rerenderWithProps({
      videoSrc: 'fake/video-file.mp4',
      loop: true,
    });
    expect(videoElement).toHaveAttribute('loop');
  });
});

describe('shouldRestartOnVideoStopped', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('shouldRestartOnVideoStopped prop restarts the video when set to true', async () => {
    const { container, getByTestId } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      // shouldRestartOnVideoStopped is true by default
    });

    const videoElement = container.querySelector('video');
    const playerContainer = getByTestId('hover-video-player-container');

    // The video's initial time should be 0
    expect(videoElement.currentTime).toBe(0);
    expect(videoElement).toBePaused();

    // Quickly start the video
    fireEvent.mouseEnter(playerContainer);

    expect(videoElement.currentTime).toBe(0);
    expect(videoElement).toBeLoading();

    await act(() => getPlayPromise(videoElement, 0));

    // The video's time should now be greater than 0 because it's playing
    expect(videoElement.currentTime).toBeGreaterThan(0);
    expect(videoElement).toBePlaying();

    // Stop the video
    fireEvent.mouseLeave(playerContainer);

    expect(videoElement.currentTime).toBeGreaterThan(0);

    // Advance a sufficient amount of time for the stop attempt to complete
    act(() => jest.advanceTimersByTime(500));

    // The video time should have been paused and reset
    expect(videoElement.currentTime).toBe(0);
    expect(videoElement.pause).toHaveBeenCalledTimes(1);
    expect(videoElement).toBePaused();
  });

  test('shouldRestartOnVideoStopped prop does not restart the video when set to false', async () => {
    const { container, getByTestId } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      shouldRestartOnVideoStopped: false,
    });

    const videoElement = container.querySelector('video');
    const playerContainer = getByTestId('hover-video-player-container');

    // The video's initial time should be 0
    expect(videoElement.currentTime).toBe(0);
    expect(videoElement).toBePaused();

    // Quickly start the video
    fireEvent.mouseEnter(playerContainer);

    expect(videoElement.currentTime).toBe(0);
    expect(videoElement).toBeLoading();

    // Wait for the promise returned by play() to resolve
    await act(() => getPlayPromise(videoElement, 0));

    // The video's time should now be greater than 0 because it's playing
    expect(videoElement.currentTime).toBeGreaterThan(0);
    expect(videoElement).toBePlaying();

    // Stop the video
    fireEvent.mouseLeave(playerContainer);

    // Advance a sufficient amount of time for the stop attempt to complete
    act(() => jest.advanceTimersByTime(500));

    // The video time should not have been reset to 0
    expect(videoElement.currentTime).toBeGreaterThan(0);
    expect(videoElement).toBePaused();
  });
});

describe('pausedOverlay and loadingOverlay', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('pausedOverlay and loadingOverlay are shown and hidden correctly as the video is started and stopped', async () => {
    const { container, getByTestId } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      loadingOverlay: <div />,
      overlayFadeTransitionDuration: 500,
      loadingStateTimeoutDuration: 500,
    });

    const videoElement = container.querySelector('video');

    const playerContainer = getByTestId('hover-video-player-container');
    const pausedOverlayWrapper = getByTestId('paused-overlay-wrapper');
    const loadingOverlayWrapper = getByTestId('loading-overlay-wrapper');

    // The paused overlay should be visible since we're in the initial idle paused state
    expect(pausedOverlayWrapper.style.opacity).toBe('1');
    expect(pausedOverlayWrapper.style.transition).toBe('opacity 500ms');
    // The loading overlay should be hidden
    expect(loadingOverlayWrapper.style.opacity).toBe('0');
    expect(loadingOverlayWrapper.style.transition).toBe('opacity 500ms');

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    // The paused overlay should still be visible while we wait for the video to play
    expect(pausedOverlayWrapper.style.opacity).toBe('1');
    // The loading overlay should still be hidden until the loading state timeout has passed
    expect(loadingOverlayWrapper.style.opacity).toBe('0');

    act(() => jest.advanceTimersByTime(500));

    // The loading state timeout has now passed without the video playing so the loading overlay should be visible
    expect(loadingOverlayWrapper.style.opacity).toBe('1');

    // Wait until the play promise has resolved
    await act(() => getPlayPromise(videoElement, 0));

    // Both overlays should be hidden now that we are playing
    expect(pausedOverlayWrapper.style.opacity).toBe('0');
    expect(loadingOverlayWrapper.style.opacity).toBe('0');

    // Fire a mouseLeave event to stop the video
    fireEvent.mouseLeave(playerContainer);

    // Advance 500ms to let the paused overlay fade back in
    act(() => jest.advanceTimersByTime(500));

    // Since we're stopping, the paused overlay should be visible again
    expect(pausedOverlayWrapper.style.opacity).toBe('1');
    // The loading overlay should still be hidden
    expect(loadingOverlayWrapper.style.opacity).toBe('0');
  });

  test('the loading state overlay is not shown if the video plays before the loading state timeout completes', async () => {
    const { container, getByTestId } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      loadingOverlay: <div />,
      loadingStateTimeoutDuration: 500,
    });

    const videoElement = container.querySelector('video');
    const playerContainer = getByTestId('hover-video-player-container');
    const loadingOverlayWrapper = getByTestId('loading-overlay-wrapper');

    // The loading overlay should be hidden initially
    expect(loadingOverlayWrapper.style.opacity).toBe('0');

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    // The loading overlay should be hidden until the loading state timeout has completed
    expect(loadingOverlayWrapper.style.opacity).toBe('0');

    act(() => jest.advanceTimersByTime(499));

    // We're 1 ms short of the loading state timeout so the loading overlay should still be hidden
    expect(loadingOverlayWrapper.style.opacity).toBe('0');

    // Resolve the play promise
    await act(() => getPlayPromise(videoElement, 0));

    // The loading overlay should stil be hidden
    expect(loadingOverlayWrapper.style.opacity).toBe('0');

    // Advance the timer sufficiently to prove the loading state timeout was cancelled
    act(() => jest.advanceTimersByTime(500));

    // The loading overlay should still be hidden
    expect(loadingOverlayWrapper.style.opacity).toBe('0');
  });
});

describe('isFocused', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('isFocused prop starts and stops the video correctly', async () => {
    const { container, rerenderWithProps } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      // isFocused is false by default
    });

    const videoElement = container.querySelector('video');

    expect(videoElement.play).toHaveBeenCalledTimes(0);
    expect(videoElement).toBePaused();

    // Set isFocused to true to start playing the video
    rerenderWithProps({ videoSrc: 'fake/video-file.mp4', isFocused: true });

    expect(videoElement.play).toHaveBeenCalledTimes(1);
    expect(videoElement).toBeLoading();

    await act(() => getPlayPromise(videoElement, 0));

    expect(videoElement).toBePlaying();

    // Set isFocused back to false to stop playing
    rerenderWithProps({ videoSrc: 'fake/video-file.mp4', isFocused: false });

    expect(videoElement.pause).toHaveBeenCalledTimes(0);
    expect(videoElement).toBePlaying();

    // Advance timers by sufficient amount of time to complete the stop attempt
    act(() => jest.advanceTimersByTime(500));

    expect(videoElement.pause).toHaveBeenCalledTimes(1);
    expect(videoElement).toBePaused();
  });

  test('other events which would normally stop the video are ignored if isFocused prop is true', async () => {
    const {
      container,
      getByTestId,
      rerenderWithProps,
    } = renderHoverVideoPlayer({ videoSrc: 'fake/video-file.mp4' });

    const playerContainer = getByTestId('hover-video-player-container');
    const videoElement = container.querySelector('video');

    expect(videoElement).toBePaused();

    // Re-render with isFocused set to true to start playing
    rerenderWithProps({ videoSrc: 'fake/video-file.mp4', isFocused: true });

    expect(videoElement.play).toHaveBeenCalledTimes(1);
    expect(videoElement).toBeLoading();

    await act(() => getPlayPromise(videoElement, 0));

    expect(videoElement).toBePlaying();

    fireEvent.mouseEnter(playerContainer);

    // Play shouldn't have been called again since we're already playing
    expect(videoElement.play).toHaveBeenCalledTimes(1);
    expect(videoElement).toBePlaying();

    expect(videoElement.pause).toHaveBeenCalledTimes(0);

    // Fire some events that would normally pause the video
    fireEvent.mouseLeave(playerContainer);
    act(() => jest.advanceTimersByTime(500));

    // The video still shouldn't have been paused
    expect(videoElement.pause).toHaveBeenCalledTimes(0);
    expect(videoElement).toBePlaying();

    fireEvent.touchStart(document.body);
    act(() => jest.advanceTimersByTime(500));

    expect(videoElement.pause).toHaveBeenCalledTimes(0);
    expect(videoElement).toBePlaying();
  });
});

describe('overlayFadeTransitionDuration', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('Stop attempts take the amount of time set by overlayFadeTransitionDuration prop if a pausedOverlay is provided', async () => {
    const { container, getByTestId } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      overlayFadeTransitionDuration: 900,
    });

    const videoElement = container.querySelector('video');

    const playerContainer = getByTestId('hover-video-player-container');

    expect(videoElement).toBePaused();

    fireEvent.mouseEnter(playerContainer);

    expect(videoElement.play).toHaveBeenCalledTimes(1);
    expect(videoElement).toBeLoading();

    await act(() => getPlayPromise(videoElement, 0));

    expect(videoElement).toBePlaying();
    fireEvent.mouseLeave(playerContainer);

    // A stop attempt should be in progress but not completed yet
    expect(videoElement.pause).toHaveBeenCalledTimes(0);
    expect(videoElement).toBePlaying();

    // Advance timer just up to before the pause timeout should complete
    jest.advanceTimersByTime(899);

    // The pause timeout should still be in progress since we're just 1ms short of the transition duration
    expect(videoElement.pause).toHaveBeenCalledTimes(0);
    expect(videoElement).toBePlaying();

    act(() => jest.advanceTimersByTime(1));

    // The stop attempt should be completed now that the full fade duration is complete
    expect(videoElement.pause).toHaveBeenCalledTimes(1);
    expect(videoElement).toBePaused();
  });

  test('Stop attempts stop immediately if a pausedOverlay is not provided', async () => {
    const { container, getByTestId } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      overlayFadeTransitionDuration: 900,
    });

    const playerContainer = getByTestId('hover-video-player-container');
    const videoElement = container.querySelector('video');

    expect(videoElement).toBePaused();
    fireEvent.mouseEnter(playerContainer);
    expect(videoElement).toBeLoading();
    await act(() => getPlayPromise(videoElement, 0));
    expect(videoElement).toBePlaying();

    fireEvent.mouseLeave(playerContainer);

    // A stop attempt should be in progress but not completed yet
    expect(videoElement.pause).toHaveBeenCalledTimes(0);
    expect(videoElement).toBePlaying();

    // Just update the timers - the pause timeout should have a timeout of 0ms and will fire
    act(() => jest.advanceTimersByTime(0));

    expect(videoElement.pause).toHaveBeenCalledTimes(1);
    expect(videoElement).toBePaused();
  });
});

describe('shouldUseOverlayDimensions', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('shouldUseOverlayDimensions prop applies the correct styling when set to true alongside a paused overlay', () => {
    const { container, getByTestId } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      // shouldUseOverlayDimensions is true by default
    });

    const pausedOverlayWrapper = getByTestId('paused-overlay-wrapper');
    const videoElement = container.querySelector('video');

    expect(videoElement).toHaveAttribute('preload', 'none');
    expect(pausedOverlayWrapper).toHaveStyleRule('position', 'relative');
    expect(videoElement).toHaveStyleRule('position', 'absolute');
  });

  test('shouldUseOverlayDimensions prop applies the correct styling when set to false alongside a paused overlay', () => {
    const { container, getByTestId } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      shouldUseOverlayDimensions: false,
    });

    const pausedOverlayWrapper = getByTestId('paused-overlay-wrapper');
    const videoElement = container.querySelector('video');

    expect(videoElement).toHaveAttribute('preload', 'metadata');
    expect(pausedOverlayWrapper).toHaveStyleRule('position', 'absolute');
    // The video element shouldn't have a position style rule set
    expect(videoElement).not.toHaveStyleRule('position', 'absolute');
  });

  test('shouldUseOverlayDimensions should be ignored and applies the correct styling when no paused overlay is provided', () => {
    const { container, queryByTestId } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      shouldUseOverlayDimensions: true,
    });

    const pausedOverlayWrapper = queryByTestId('paused-overlay-wrapper');
    const videoElement = container.querySelector('video');

    expect(videoElement).toHaveAttribute('preload', 'metadata');
    expect(pausedOverlayWrapper).not.toBeInTheDocument();
    // The video element shouldn't have a position style rule set
    expect(videoElement).not.toHaveStyleRule('position');
  });
});
