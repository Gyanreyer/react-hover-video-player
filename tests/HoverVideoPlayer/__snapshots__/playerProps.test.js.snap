// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video props isVideoMuted prop correctly sets muted attribute on video 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Video props shouldVideoLoop prop correctly sets loop attribute on video 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`isFocused isFocused prop starts and stops the video correctly 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`isFocused other events which would normally stop the video are ignored if isFocused prop is true 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`overlayFadeTransitionDuration Stop attempts stop immediately if a pausedOverlay is not provided 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`overlayFadeTransitionDuration Stop attempts take the amount of time set by overlayFadeTransitionDuration prop if a pausedOverlay is provided 1`] = `
.emotion-2 {
  position: relative;
}

.emotion-0 {
  position: relative;
  display: block;
  width: 100%;
  z-index: 1;
  pointer-events: none;
}

.emotion-0 > * {
  display: block;
  width: 100%;
}

.emotion-1 {
  width: 100%;
  display: block;
  object-fit: cover;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="emotion-2"
    data-testid="hover-video-player-container"
  >
    <div
      class="emotion-0"
      data-testid="paused-overlay-wrapper"
      style="opacity: 1; transition: opacity 900ms;"
    >
      <div />
    </div>
    <video
      class="emotion-1"
      loop=""
      playsinline=""
      preload="none"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`pausedOverlay and loadingOverlay pausedOverlay and loadingOverlay are shown and hidden correctly as the video is started and stopped 1`] = `
.emotion-3 {
  position: relative;
}

.emotion-0 {
  position: relative;
  display: block;
  width: 100%;
  z-index: 1;
  pointer-events: none;
}

.emotion-0 > * {
  display: block;
  width: 100%;
}

.emotion-1 {
  z-index: 2;
  pointer-events: none;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

.emotion-2 {
  width: 100%;
  display: block;
  object-fit: cover;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="emotion-3"
    data-testid="hover-video-player-container"
  >
    <div
      class="emotion-0"
      data-testid="paused-overlay-wrapper"
      style="opacity: 1; transition: opacity 500ms;"
    >
      <div />
    </div>
    <div
      class="emotion-1"
      data-testid="loading-overlay-wrapper"
      style="opacity: 0; transition: opacity 500ms;"
    >
      <div />
    </div>
    <video
      class="emotion-2"
      loop=""
      playsinline=""
      preload="none"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`pausedOverlay and loadingOverlay the loading state overlay is not shown if the video plays before the loading state timeout completes 1`] = `
.emotion-2 {
  position: relative;
}

.emotion-1 {
  width: 100%;
  display: block;
  object-fit: cover;
}

.emotion-0 {
  z-index: 2;
  pointer-events: none;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="emotion-2"
    data-testid="hover-video-player-container"
  >
    <div
      class="emotion-0"
      data-testid="loading-overlay-wrapper"
      style="opacity: 0; transition: opacity 400ms;"
    >
      <div />
    </div>
    <video
      class="emotion-1"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`shouldRestartOnVideoStopped shouldRestartOnVideoStopped prop does not restart the video when set to false 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`shouldRestartOnVideoStopped shouldRestartOnVideoStopped prop restarts the video when set to true 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`shouldUseOverlayDimensions shouldUseOverlayDimensions prop applies the correct styling when set to false alongside a paused overlay 1`] = `
.emotion-2 {
  position: relative;
}

.emotion-1 {
  width: 100%;
  display: block;
  object-fit: cover;
}

.emotion-0 {
  position: relative;
  display: block;
  width: 100%;
  z-index: 1;
  pointer-events: none;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

.emotion-0 > * {
  display: block;
  width: 100%;
}

<div>
  <div
    class="emotion-2"
    data-testid="hover-video-player-container"
  >
    <div
      class="emotion-0"
      data-testid="paused-overlay-wrapper"
      style="opacity: 1; transition: opacity 400ms;"
    >
      <div />
    </div>
    <video
      class="emotion-1"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`shouldUseOverlayDimensions shouldUseOverlayDimensions prop applies the correct styling when set to true alongside a paused overlay 1`] = `
.emotion-2 {
  position: relative;
}

.emotion-0 {
  position: relative;
  display: block;
  width: 100%;
  z-index: 1;
  pointer-events: none;
}

.emotion-0 > * {
  display: block;
  width: 100%;
}

.emotion-1 {
  width: 100%;
  display: block;
  object-fit: cover;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

<div>
  <div
    class="emotion-2"
    data-testid="hover-video-player-container"
  >
    <div
      class="emotion-0"
      data-testid="paused-overlay-wrapper"
      style="opacity: 1; transition: opacity 400ms;"
    >
      <div />
    </div>
    <video
      class="emotion-1"
      loop=""
      playsinline=""
      preload="none"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`shouldUseOverlayDimensions shouldUseOverlayDimensions should be ignored and applies the correct styling when no paused overlay is provided 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;
